plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterManagerImpl.java

```diff
@@ -845,30 +865,87 @@ private void validateKubernetesClusterScaleParameters(ScaleKubernetesClusterCmd
-        }
-
         final KubernetesSupportedVersion clusterVersion = kubernetesSupportedVersionDao.findById(kubernetesCluster.getKubernetesVersionId());
         if (clusterVersion == null) {
             throw new CloudRuntimeException(String.format("Invalid Kubernetes version associated with Kubernetes cluster : %s", kubernetesCluster.getName()));
         }
 
+        if (!(kubernetesCluster.getState().equals(KubernetesCluster.State.Created) ||
+                kubernetesCluster.getState().equals(KubernetesCluster.State.Running) ||
+                kubernetesCluster.getState().equals(KubernetesCluster.State.Stopped))) {
+            throw new PermissionDeniedException(String.format("Kubernetes cluster %s is in %s state and can not be scaled", kubernetesCluster.getName(), kubernetesCluster.getState().toString()));
+        }
+
+        if (isAutoscalingEnabled != null && isAutoscalingEnabled) {
+            if (clusterSize != null || serviceOfferingId != null || nodeIds != null) {
+                throw new InvalidParameterValueException("autoscaling can not be passed along with nodeids or clustersize or service offering");
+            }
+
+            if (!KubernetesVersionManagerImpl.versionSupportsAutoscaling(clusterVersion)) {
+                throw new InvalidParameterValueException(String.format("Autoscaling requires Kubernetes Version %s or above",
+                    KubernetesVersionManagerImpl.MINIMUN_AUTOSCALER_SUPPORTED_VERSION ));
+            }
+
+            validateEndpointUrl();
+
+            if (minSize == null || maxSize == null) {
+                throw new InvalidParameterValueException("autoscaling requires minsize and maxsize to be passed");
+            }
+            if (minSize < 1) {
+                throw new InvalidParameterValueException("minsize must be at least than 1");
+            }
+            if (maxSize <= minSize) {
+                throw new InvalidParameterValueException("maxsize must be greater than or equal to minsize");
+            }
+        }
+
+        if (nodeIds != null) {
+            if (clusterSize != null || serviceOfferingId != null) {
+                throw new InvalidParameterValueException("nodeids can not be passed along with clustersize or service offering");
+            }
+            List<KubernetesClusterVmMapVO> nodes = kubernetesClusterVmMapDao.listByClusterIdAndVmIdsIn(kubernetesCluster.getId(), nodeIds);
+            // Do all the nodes exist ?
+            if (nodes == null || nodes.size() != nodeIds.size()) {
+                throw new InvalidParameterValueException("Invalid node ids");
+            }
+            // Ensure there's always a master
+            long mastersToRemove = nodes.stream().filter(x -> x.isMaster()).count();
+            if (mastersToRemove >= kubernetesCluster.getMasterNodeCount()) {
+                throw new InvalidParameterValueException("Can not remove all masters from a cluster");
+            }
```

Comment:
Are we allowing scaling for master nodes?

Reply:
The autoscaler will not remove master nodes, however, nodeids can be passed, ensuring that users do not accidentally delete master nodes